cmd/ecr-credential-provider/main.go

```diff
@@ -0,0 +1,167 @@
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+
+	plugin "k8s.io/cloud-provider-aws/cmd/ecr-credential-provider/framework"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubelet/pkg/apis/credentialprovider/v1alpha1"
+)
+
+var ecrPattern = regexp.MustCompile(`^(\d{12})\.dkr\.ecr(\-fips)?\.([a-zA-Z0-9][a-zA-Z0-9-_]*)\.(amazonaws\.com(\.cn)?|sc2s\.sgov\.gov|c2s\.ic\.gov)$`)
+
+// ECR abstracts the calls we make to aws-sdk for testing purposes
+type ECR interface {
+	GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error)
+}
+
+// ECRProvider is responsible for instantiating a new ECR instance
+type ECRProvider interface {
+	NewECR(region string, registryID string) (*ecr.ECR, error)
+}
+
+type ecrPlugin struct {
+	ecrProvider ECRProvider
+	ecr         ECR
+}
+
+func defaultECRProvider(region string, registryID string) (*ecr.ECR, error) {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String(region)},
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return ecr.New(sess), nil
+}
+
+func (e *ecrPlugin) GetCredentials(ctx context.Context, image string, args []string) (*v1alpha1.CredentialProviderResponse, error) {
+	registryID, region, registry, err := parseRepoURL(image)
+	if err != nil {
+		return nil, err
+	}
+
+	if e.ecr == nil {
+		e.ecr, err = defaultECRProvider(region, registryID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	output, err := e.ecr.GetAuthorizationToken(&ecr.GetAuthorizationTokenInput{
+		RegistryIds: []*string{aws.String(registryID)},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if output == nil {
+		return nil, errors.New("response output from ECR was nil")
+	}
+
+	if len(output.AuthorizationData) == 0 {
+		return nil, errors.New("authorization data was empty")
+	}
+
+	data := output.AuthorizationData[0]
+	if data.AuthorizationToken == nil {
+		return nil, errors.New("authorization token in response was nil")
+	}
+
+	decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
+	if err != nil {
+		return nil, err
+	}
+
+	parts := strings.SplitN(string(decodedToken), ":", 2)
+	if len(parts) != 2 {
+		return nil, errors.New("error parsing username and password from authorization token")
+	}
+
+	var cacheDuration *metav1.Duration
+	expiresAt := data.ExpiresAt
+	if expiresAt == nil {
+		// explicitly set cache duration to 0 if expiresAt was nil so that
+		// kubelet does not cache it in-memory
+		cacheDuration = &metav1.Duration{Duration: 0}
+	} else {
+		// substracting two hours in order to compensate for the time loss between
+		// the token creation and passing it all the way to kubelet.
+		duration := expiresAt.Sub(time.Now()) - (2 * time.Hour)
```

Comment:
2 hours seems sufficient, but I wonder if subtraction should be based on some percentage rather than absolute values. For example: "half the expiration" vs "expiration - 2 hours". This way if the expiration was for whatever reason less than 2 hours, it would still be valid.

Reply:
Sounds fair. ECR docs say it's always 12 hours, but better safe than sorry. Will update.